File: backend/src/tools/_hooks/advisor_approval.py
"""Pre-hook that gates main-agent terminal submissions on an advisor approval.

The hook scans ``context.conversation_messages`` for the most recent
``ask_advisor`` result, pairs it with its originating ``ToolUseBlock`` to
recover the ``tool_name`` argument, and rejects the terminal call unless
the advisor approved THIS specific terminal.

Wiring is per-terminal: each gated tool's ``@tool`` decorator carries
``pre_hooks=(AdvisorApprovalPreHook("<own_name>"),)``. Helper / subagent
terminals (``submit_advisor_feedback``, ``submit_exploration_result``)
intentionally omit the hook.
"""

from __future__ import annotations

from pydantic import BaseModel

from message.message import Message, ToolResultBlock, ToolUseBlock
from tools._framework.core.context import ToolExecutionContextService
from tools._framework.core.hooks import HookResult


_ADVISOR_HELPER_ROLE = "advisor"
_VALID_VERDICTS = frozenset({"approve", "reject"})

# Single prescriptive message for every fail branch. Agents already have the
# prior advisor exchange (reject summaries, prior tool selection, errored
# ask_advisor results) in their own conversation history, so echoing branch-
# specific detail back would duplicate context the agent already sees. Keep
# the user-facing message uniform; per-branch distinction is preserved in
# ``HookResult.metadata["reason"]`` for ops/debugging.
_MSG_BLOCKED = (
    "BLOCKED: You must get approval from advisor before submitting this "
    "terminal. Call ask_advisor(tool_name=\"{tool}\", tool_payload=...) and "
    "resubmit only after the advisor returns verdict=\"approve\"."
)


class AdvisorApprovalPreHook:
    """Per-terminal hook: requires advisor approval for THIS tool.

    The hook is instance-per-terminal so ``target_tool`` matches the
    decorator's ``name``. ``validate_hook_targets`` reads ``target_tool``
    via ``getattr``; an instance attribute is sufficient.
    """

    def __init__(self, target_tool: str) -> None:
        self.target_tool = target_tool
        self.name = f"advisor_approval:{target_tool}"

    async def run(
        self,
        tool_input: BaseModel,
        context: ToolExecutionContextService,
    ) -> HookResult[BaseModel]:
        messages = context.get("conversation_messages") or []
        result_block, originating = _find_latest_advisor_pair(messages)
        reason = self._classify(result_block, originating)
        if reason is None:
            return HookResult.pass_(tool_input)
        return HookResult.fail(
            _MSG_BLOCKED.format(tool=self.target_tool),
            metadata={"policy": "advisor_approval", "reason": reason},
        )

    def _classify(
        self,
        result_block: ToolResultBlock | None,
        originating: ToolUseBlock | None,
    ) -> str | None:
        """Return a failure-reason tag, or ``None`` if the gate passes.

        The tag drives ops/observability via ``HookResult.metadata["reason"]``;
        the agent-facing message is uniform across branches.
        """
        if result_block is None:
            return "missing"
        if result_block.is_error:
            return "advisor_failed"
        verdict = result_block.metadata.get("verdict")
        if verdict not in _VALID_VERDICTS:
            return "structural"
        if verdict == "reject":
            return "rejected"
        if originating is None:
            return "unpaired"
        if originating.input.get("tool_name") != self.target_tool:
            return "wrong_tool"
        return None


def _find_latest_advisor_pair(
    messages: list[Message],
) -> tuple[ToolResultBlock | None, ToolUseBlock | None]:
    """Reverse-walk for the most recent advisor result and forward-pair it."""
    for msg in reversed(messages):
        if msg.role != "user":
            continue
        for block in reversed(msg.content):
            if (
                isinstance(block, ToolResultBlock)
                and block.metadata.get("helper_role") == _ADVISOR_HELPER_ROLE
            ):
                originating = _find_originating_ask_advisor(messages, block.tool_use_id)
                return block, originating
    return None, None


def _find_originating_ask_advisor(
    messages: list[Message],
    tool_use_id: str,
) -> ToolUseBlock | None:
    for msg in messages:
        if msg.role != "assistant":
            continue
        for block in msg.tool_uses:
            if block.tool_use_id == tool_use_id and block.name == "ask_advisor":
                return block
    return None


__all__ = ["AdvisorApprovalPreHook"]

File: agent-core/crates/eos-engine/src/tool_call/hooks/advisor_approval.rs
//! The advisor-approval pre-hook — a **stateless** gate that infers the verdict
//! from the conversation transcript.
//!
//! There is no port and no engine/agent state: the verdict exists only as a
//! `submit_advisor_outcome` result block in the transcript, surfaced as the
//! `ask_advisor` result. The gate reverse-walks [`ExecutionMetadata::conversation`]
//! for the latest advisor result, pairs it to the originating `ask_advisor`
//! tool-use, and classifies — re-deriving the decision on demand, never reading a
//! cached verdict (advisor remediation plan §2b / §9).

use eos_llm_client::{ContentBlock, Message, MessageRole};
use eos_types::ToolUseId;

use eos_tool::{ExecutionMetadata, ToolError, ToolName};

use super::{HookDenial, HookOutcome};

const ADVISOR_HELPER_ROLE: &str = "advisor";
const VALID_VERDICTS: [&str; 2] = ["approve", "reject"];

const ADVISOR_APPROVAL_MESSAGE_PREFIX: &str =
    "BLOCKED: You must get approval from advisor before submitting this terminal. \
     Call ask_advisor(tool_name=\"";
const ADVISOR_APPROVAL_MESSAGE_SUFFIX: &str =
    "\", tool_payload=...) and resubmit only after the advisor returns verdict=\"approve\".";

fn blocked_message(tool: ToolName) -> String {
    format!(
        "{ADVISOR_APPROVAL_MESSAGE_PREFIX}{}{ADVISOR_APPROVAL_MESSAGE_SUFFIX}",
        tool.as_str()
    )
}

/// `AdvisorApprovalPreHook.run`: scan the transcript, classify, deny or pass. The
/// missing-conversation case (no advisor result) classifies `missing` and denies,
/// matching Rust's "no conversation → reason `missing`".
pub(crate) async fn run_advisor_approval(
    tool: ToolName,
    ctx: &ExecutionMetadata,
) -> Result<HookOutcome, ToolError> {
    let (result, originating) = find_latest_advisor_pair(&ctx.conversation);
    match classify_advisor_approval(result, originating, tool) {
        None => Ok(HookOutcome::pass()),
        Some(reason) => Ok(HookOutcome::Deny(
            HookDenial::new(blocked_message(tool), "advisor_approval").with_reason(reason),
        )),
    }
}

/// Reverse-walk `User` messages for the latest `tool_result` block whose metadata
/// carries `helper_role == "advisor"`; pair it with the originating `ask_advisor`
/// tool-use (matched by `tool_use_id`). Port of `_find_latest_advisor_pair`.
fn find_latest_advisor_pair(
    messages: &[Message],
) -> (Option<&ContentBlock>, Option<&ContentBlock>) {
    for msg in messages.iter().rev() {
        if msg.role != MessageRole::User {
            continue;
        }
        for block in msg.content.iter().rev() {
            if let ContentBlock::ToolResult {
                tool_use_id,
                metadata,
                ..
            } = block
            {
                if metadata.get("helper_role").and_then(|v| v.as_str()) == Some(ADVISOR_HELPER_ROLE)
                {
                    let originating = find_originating_ask_advisor(messages, tool_use_id);
                    return (Some(block), originating);
                }
            }
        }
    }
    (None, None)
}

/// Forward-walk `Assistant` messages for the `ask_advisor` tool-use with the given
/// id. Port of `_find_originating_ask_advisor`.
fn find_originating_ask_advisor<'a>(
    messages: &'a [Message],
    tool_use_id: &ToolUseId,
) -> Option<&'a ContentBlock> {
    for msg in messages {
        if msg.role != MessageRole::Assistant {
            continue;
        }
        for block in &msg.content {
            if let ContentBlock::ToolUse {
                tool_use_id: id,
                name,
                ..
            } = block
            {
                if id == tool_use_id && name == "ask_advisor" {
                    return Some(block);
                }
            }
        }
    }
    None
}

/// Classify the advisor pair against `target_tool`. `Some(tag)` denies; `None`
/// passes. Order (verbatim `_classify`): `missing` → `advisor_failed` →
/// `structural` → `rejected` → `unpaired` → `wrong_tool` → pass.
fn classify_advisor_approval(
    result: Option<&ContentBlock>,
    originating: Option<&ContentBlock>,
    target_tool: ToolName,
) -> Option<&'static str> {
    let Some(ContentBlock::ToolResult {
        is_error, metadata, ..
    }) = result
    else {
        return Some("missing");
    };
    if *is_error {
        return Some("advisor_failed");
    }
    let Some(verdict) = metadata
        .get("verdict")
        .and_then(|v| v.as_str())
        .filter(|v| VALID_VERDICTS.contains(v))
    else {
        return Some("structural");
    };
    if verdict == "reject" {
        return Some("rejected");
    }
    // verdict == "approve" — require the originating ask_advisor to target this terminal.
    let Some(ContentBlock::ToolUse { input, .. }) = originating else {
        return Some("unpaired");
    };
    if input.get("tool_name").and_then(|v| v.as_str()) != Some(target_tool.as_str()) {
        return Some("wrong_tool");
    }
    None
}

#[cfg(test)]
#[path = "../../../tests/tool_call/hooks/advisor_approval/mod.rs"]
mod tests;
